### ISSUE TITLE
Settings lost if system is force-shutdown

### ISSUE THREAD

**Issue by darksylinc** [git-cola/git-cola#1241]
On Xubuntu 20.04 LTS settings will often (but not always) get wiped out if I go to `Whisker Menu -> Shutdown -> Shutdown` or just press the Power Button on the PC (i.e. ACPI shutdown) while git-cola is still open.

It's annoying because I lose everything: Favorites, UI layout, saved repos, everything.

A quick look at [settings.py](https://github.com/git-cola/git-cola/blob/38c80ca52aee0f00c0f8d93e62dff2f893eb22a4/cola/settings.py#L148) reveals the why settings are destroyed: it's not a safe method of writing config files.

The safe way to save settings (or any file) is:

## Save

1. Write new settings.tmp
2. Rename `settings` to `settings.bak`
3. Rename `settings.tmp` to `settings`
4. Flush disk
5. Delete `settings.bak`

## Load

1. Load `settings.bak`. If it suceeds, we're done
2. Load `settings`. If it suceeds, we're done
3. Load settings `settings.tmp`, this is a bit dangerous without CRC check though, as the file may be corrupted
4. If we're here, either settings were lost or this is a new installation

## Flushing disk

In C to flush the disk one would do:

```cpp
int status = fflush( m_handle );
if( status != 0 )
	m_statusBits |= fstream::badbit;

int fileDescriptor = fileno( m_handle );

if( fileDescriptor == -1 )
{
	m_statusBits |= fstream::badbit;
	return;
}

if( APPLE )
{
	status = fcntl( fileDescriptor, F_FULLFSYNC, NULL );
	if( status )
	{
		// If we are not on a file system that supports this,
		// then fall back to a plain fsync.
		status = ::fsync( fileDescriptor );
		if( status != 0 )
			m_statusBits |= fstream::badbit;
	}
}
else if( NOT WINDOWS )
{
	status = ::fdatasync( fileDescriptor );

	if( status != 0 )
		m_statusBits |= fstream::badbit;
}
else
{
	status = _commit( fileDescriptor );
	if( status != 0 )
		m_statusBits |= fstream::badbit;
}
```

I dunno if Python gives access to file descriptors or if it already has secure save/open utilities to take care of this algorithm automatically.

Cheers.

**Comment by davvid** [comment]
Thanks for the heads-up. Indeed, the temp-file atomic-rename swap would be a good approach to take in general. We don't currently flush the disk but that would  be worth doing.

Python does provide access to `os.sync()` so we should at least use it. It's a much heavier hammer than a C fdatasync chisel, but then again it is python.

**Comment by davvid** [comment]
Thanks again for the detailed explanation. This approach has been implemented and will be included in the upcoming v4.0.3 release which will be tagged soon.

I focused almost exclusively on usability and implementation fixes so this one seemed worth including since it definitely makes things more robust.

**Comment by darksylinc** [comment]
Thank you!!!

While it's always possible that the algorithm could be missing a step to avoid data loss for all sort of scenarios (database developers have covered this issue in great length; every OS and filesystem has its own nuisances), this should fix the most common of data loss causes, which right now is happening with enough frequency to be a problem.

Thank you!
